File: output/csbc_semish_2026/x_ai_grok_code_fast_1/few/python/Fila_2015.py
import sys
import random

sys.setrecursionlimit(200000)

class Node:
    def __init__(self, h):
        self.h = h
        self.prior = random.randint(0, 1 << 60)
        self.size = 1
        self.max_h = h
        self.left = None
        self.right = None

def size(node):
    return node.size if node else 0

def update(node):
    if not node:
        return
    node.size = 1 + size(node.left) + size(node.right)
    node.max_h = node.h
    if node.left:
        node.max_h = max(node.max_h, node.left.max_h)
    if node.right:
        node.max_h = max(node.max_h, node.right.max_h)

def split(node, k):
    if not node:
        return None, None
    if size(node.left) >= k:
        left, node.left = split(node.left, k)
        update(node)
        return left, node
    else:
        node.right, right = split(node.right, k - size(node.left) - 1)
        update(node)
        return node, right

def merge(left, right):
    if not left:
        return right
    if not right:
        return left
    if left.prior > right.prior:
        left.right = merge(left.right, right)
        update(left)
        return left
    else:
        right.left = merge(left, right.left)
        update(right)
        return right

def get_kth(node, k):
    if not node:
        return None
    if size(node.left) >= k:
        return get_kth(node.left, k)
    if size(node.left) + 1 == k:
        return node
    return get_kth(node.right, k - size(node.left) - 1)

def find_rightmost(node, val, base):
    if not node or node.max_h <= val:
        return 0
    if node.right and node.right.max_h > val:
        return find_rightmost(node.right, val, base + size(node.left) + 1)
    if node.h > val:
        return base + size(node.left) + 1
    return find_rightmost(node.left, val, base)

input_data = sys.stdin.read().split()
idx = 0
N = int(input_data[idx])
idx += 1
heights = []
for i in range(N):
    heights.append(int(input_data[idx]))
    idx += 1
Q = int(input_data[idx])
idx += 1
root = None
for h in heights:
    new_node = Node(h)
    root = merge(root, new_node)
for _ in range(Q):
    T = int(input_data[idx])
    I = int(input_data[idx + 1])
    X = int(input_data[idx + 2])
    idx += 3
    if T == 0:
        pos = I + 1 if I > 0 else 1
        left, right = split(root, pos - 1)
        new_node = Node(X)
        root = merge(merge(left, new_node), right)
    else:
        node = get_kth(root, I)
        val = node.h + X
        left, right = split(root, I - 1)
        pos = find_rightmost(left, val, 0)
        root = merge(left, right)
        print(pos)

File: output/csbc_semish_2026/x_ai_grok_code_fast_1/few/cpp/Fila_2015.cpp

#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

struct Node {
    ll pri, alt, max_alt;
    int size;
    Node *l, *r;
    Node(ll a) : pri(rand()), alt(a), max_alt(a), size(1), l(nullptr), r(nullptr) {}
};

int get_size(Node* t) { return t ? t->size : 0; }

void update(Node* t) {
    if (!t) return;
    t->size = 1 + get_size(t->l) + get_size(t->r);
    t->max_alt = t->alt;
    if (t->l) t->max_alt = max(t->max_alt, t->l->max_alt);
    if (t->r) t->max_alt = max(t->max_alt, t->r->max_alt);
}

void split(Node* t, int k, Node*& a, Node*& b) {
    if (!t) { a = b = nullptr; return; }
    if (get_size(t->l) >= k) {
        split(t->l, k, a, t->l);
        b = t;
    } else {
        split(t->r, k - get_size(t->l) - 1, t->r, b);
        a = t;
    }
    update(t);
}

Node* merge(Node* a, Node* b) {
    if (!a) return b;
    if (!b) return a;
    if (a->pri > b->pri) {
        a->r = merge(a->r, b);
        update(a);
        return a;
    } else {
        b->l = merge(a, b->l);
        update(b);
        return b;
    }
}

Node* get_kth(Node* t, int k) {
    if (!t) return nullptr;
    if (get_size(t->l) + 1 == k) return t;
    if (get_size(t->l) >= k) return get_kth(t->l, k);
    return get_kth(t->r, k - get_size(t->l) - 1);
}

int find_last_greater(Node* t, ll val, int offset = 0) {
    if (!t) return -1;
    if (t->alt > val) {
        int res = find_last_greater(t->r, val, offset + get_size(t->l) + 1);
        if (res != -1) return res;
        return offset + get_size(t->l);
    } else {
        if (t->r && t->r->max_alt > val) return find_last_greater(t->r, val, offset + get_size(t->l) + 1);
        return -1;
    }
}

int main() {
    srand(time(0));
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int N;
    cin >> N;
    Node* root = nullptr;
    for (int i = 0; i < N; ++i) {
        ll a;
        cin >> a;
        Node* novo = new Node(a);
        root = merge(root, novo);
    }
    int Q;
    cin >> Q;
    for (int q = 0; q < Q; ++q) {
        int T, I;
        ll X;
        cin >> T >> I >> X;
        if (T == 0) {
            Node* novo = new Node(X);
            if (I == 0) {
                root = merge(novo, root);
            } else {
                Node *left, *right;
                split(root, I, left, right);
                root = merge(merge(left, novo), right);
            }
        } else {
            Node* person = get_kth(root, I);
            ll H = person->alt;
            ll val = H + X;
            Node *left, *right;
            split(root, I - 1, left, right);
            int rel_pos = find_last_greater(left, val);
            if (rel_pos == -1) {
                cout << 0 << '\n';
            } else {
                cout << rel_pos + 1 << '\n';
            }
            root = merge(left, right);
        }
    }
    return 0;
}
